sdk/identity/azure-identity/azure/identity/aio/_authn_client.py

```diff
@@ -65,14 +65,21 @@ def __init__(
         token = self._deserialize_and_cache_token(response=response, scopes=scopes, request_time=request_time)
         return token
 
-    async def obtain_token_by_refresh_token(self, scopes: "Iterable[str]", username: str) -> "Optional[AccessToken]":
+    async def obtain_token_by_refresh_token(self, scopes: "Iterable[str]", username: "Optional[str]") -> "Optional[AccessToken]":
         """Acquire an access token using a cached refresh token. Returns ``None`` when that fails, or the cache has no
         refresh token. This is only used by SharedTokenCacheCredential and isn't robust enough for anything else."""
 
-        # find account matching username
-        accounts = self._cache.find(TokenCache.CredentialType.ACCOUNT, query={"username": username})
+        # if an username is provided, restrect our search to accounts that have that username
+        query = {"username": username} if username else {}
+        accounts = self._cache.find(TokenCache.CredentialType.ACCOUNT, query=query)
+
+        # If there are multiple accounts in the cache and a username has not been specified, give up, since
+        # we don't know what one to pick.
+        if len(accounts) != 1 and not username:
```

Comment:
This is true when we have 2+ accounts for the same user; we ought to proceed in that case. I'd also like to defend against accounts having no username, because I don't know whether that's possible 😸

Which has me thinking:
```py
# username is required iff the cache contains accounts for multiple, or unidentified, users
if not username and len(accounts) > 1:
    usernames = {account.get("username") for account in accounts}
    if len(usernames) != 1 or not all(usernames):
        return None
```

Reply:
I looked briefly at what we do in Java and .NET here (on both the question of how the empty string is used for a username and what happens in the case where there are multiple identities in the cache with the same username).

.NET:  In .NET, an [empty string is treated as the same as `null` username](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs#L98-L103).  In this case, we .NET requires that a single identity be in the cache.  If a username is provided, the cache is filtered based on username, but [after filtering there must be a *single* account left](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs#L106).

Java: In [Java](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java#L72-L97). an empty string for the username means "find the account in the cache with an empty username".  In the case where the username is not provided, and the cache contains multiple identities, even if they have the same username, the overall call would fail (the structure of this path was a little confusing to me at first, but I believe that what ends up happening is that if there are multiple entries the `IAccount` set, on the first iteration of the [loop](https://github.com/Azure/azure-sdk-for-java/pull/5917/files#diff-7ccb32d40bee8ed1a674b839b685263fR84-R95), `requestedAccount` will be `null`, and so we'll set `cached` to the first element in the set.  Then on the next iteration, `requestedAccount` will be set, but one of the `homeAccountId()` (a unique identifier across tenets) or `environment()` will differ (even if the username is the same) and we will continue to fail.  In the case where there are multiple identities with the same username and a username was provided, we just use the "first" one from the cache.

Given the above, I'm tempted to revert back to the behavior I had previously, where we fail in all cases where there were multiple identities after doing our filtering.  I'm not 100% certain that our strategy of "just pick the first one that works" is actually correct, for example, if there are multiple `matell@microsoft.com`'s in the cache with different `HomeAccountIds` or `Environments`, it's not clear to me which of the identities we'd want to use there.  The behavior in this case would then align with .NET.

You're the owner, @chlowell and have spent more time in this space than I have, so I defer to what you want.  Happy to do what you propose here if you'd like.

Reply:
I'm not certain this strategy is exactly correct either, though when first implementing this I looked through MSAL Python for a reason not to treat accounts with the same username as fungible, and didn't find one. In the future I intend to rewrite this credential to use a higher-level MSAL API, as does our .NET library, and not use the cache directly.

But for today, as an argument for ignoring the multiplicity of accounts, allow me to describe my situation. My cache contains one refresh token, and 4 accounts differing only in their 'realm' guid. The accounts share the same home account ID with the refresh token, so it seems reasonable to treat it as belonging to any of them. The refresh token is valid. If we insist there be only one account for my username, I can't use the SDK's single sign-on feature 😞

Reply:
Okay - I'll make your proposed change!